homeassistant/components/switchbot_cloud/sensor.py

```diff
@@ -0,0 +1,83 @@
+"""Platform for sensor integration."""
+
+from switchbot_api import Device, SwitchBotAPI
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import PERCENTAGE, UnitOfTemperature
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import SwitchbotCloudData
+from .const import DOMAIN
+from .coordinator import SwitchBotCoordinator
+from .entity import SwitchBotCloudEntity
+
+SENSOR_TYPE_TEMPERATURE = "temperature"
+SENSOR_TYPE_HUMIDITY = "humidity"
+SENSOR_TYPE_BATTERY = "battery"
+
+METER_PLUS_SENSOR_DESCRIPTIONS = (
+    SensorEntityDescription(
+        key=SENSOR_TYPE_TEMPERATURE,
+        device_class=SensorDeviceClass.TEMPERATURE,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_TYPE_HUMIDITY,
+        device_class=SensorDeviceClass.HUMIDITY,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=PERCENTAGE,
+    ),
+    SensorEntityDescription(
+        key=SENSOR_TYPE_BATTERY,
+        device_class=SensorDeviceClass.BATTERY,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=PERCENTAGE,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up SwitchBot Cloud entry."""
+    data: SwitchbotCloudData = hass.data[DOMAIN][config.entry_id]
+
+    async_add_entities(
+        SwitchBotCloudSensor(data.api, device, coordinator, description)
+        for device, coordinator in data.devices.sensors
+        for description in METER_PLUS_SENSOR_DESCRIPTIONS
+    )
+
+
+class SwitchBotCloudSensor(SwitchBotCloudEntity, SensorEntity):
+    """Representation of a SwitchBot Cloud sensor entity."""
+
+    def __init__(
+        self,
+        api: SwitchBotAPI,
+        device: Device,
+        coordinator: SwitchBotCoordinator,
+        description: SensorEntityDescription,
+    ) -> None:
+        """Initialize SwitchBot Cloud sensor entity."""
+        super().__init__(api, device, coordinator)
+        self.entity_description = description
+        self._attr_unique_id = f"{device.device_id}_{description.key}"
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        if not self.coordinator.data:
+            return
```

Comment:
One thing I am wondering: When can coordinator.data be None? That would mean the coordinator update failed, but then the entity would render unavailable, right?

Reply:
I think you're right, if the code gets here then we should have data

Reply:
I take that back! I believe if the SwitchBot API request fails, we can get a `NoneType` here so we need the short circuit. It does seem to fail occasionally.

Reply:
Just tested and observed this happening.

Reply:
What error is it raising? We should catch that in the coordinator and raise `UpdateFailed`, which then renders the entities as unavailable

Reply:
A timeout error during the fetch. It does render the entities as unavailable in this case.

Reply:
Ahh ok well I think the `native_value` solution is cleaner too so let's go with that. Thanks!

Reply:
Scratch that again! We get an error on sensor entity init:

```
AttributeError: 'NoneType' object has no attribute 'get'
2024-05-09 23:35:23.188 ERROR (MainThread) [homeassistant.components.sensor] Error adding entity sensor.temp_sensor_battery for domain sensor with platform switchbot_cloud
Traceback (most recent call last):
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity_platform.py", line 580, in _async_add_entities
    await coro
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity_platform.py", line 892, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity.py", line 1366, in add_to_platform_finish
    self.async_write_ha_state()
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity.py", line 1011, in async_write_ha_state
    self._async_write_ha_state()
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity.py", line 1137, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity.py", line 1072, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/home-assistant-core/homeassistant/helpers/entity.py", line 1017, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/workspaces/home-assistant-core/homeassistant/components/sensor/__init__.py", line 543, in state
    value = self.native_value
            ^^^^^^^^^^^^^^^^^
  File "/workspaces/home-assistant-core/homeassistant/components/switchbot_cloud/sensor.py", line 81, in native_value
    return self.coordinator.data.get(self.entity_description.key)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
```

Now I'm back to thinking the short circuit is the way to go. I can see that pattern used in other integrations too.